assets/js/base/utils/render-frontend.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * External dependencies
+ */
+import { render } from 'react-dom';
+import BlockErrorBoundary from '@woocommerce/base-components/block-error-boundary';
+import { cloneElement } from '@wordpress/element';
+import parse from 'html-react-parser';
+
+/**
+ * Using defined block attributes, parse an array of attributes to get valid values, or fallback to
+ * the default value.
+ *
+ * @param {Object} blockAttributes Object containing block attributes.
+ * @param {Array} rawAttributes Dataset from DOM or an array of attributes.
+ * @return {Array} Array of parsed attributes.
+ */
+export const getValidBlockAttributes = (
+	blockAttributes,
+	rawAttributes = []
+) => {
+	const attributes = [];
+
+	Object.keys( blockAttributes ).forEach( ( key ) => {
+		if ( typeof rawAttributes[ key ] !== 'undefined' ) {
+			switch ( blockAttributes[ key ].type ) {
+				case 'boolean':
+					attributes[ key ] = rawAttributes[ key ] !== 'false';
+					break;
+				case 'number':
+					attributes[ key ] = parseInt( rawAttributes[ key ], 10 );
+					break;
+				case 'array':
+				case 'object':
+					attributes[ key ] = JSON.parse( rawAttributes[ key ] );
+					break;
+				default:
+					attributes[ key ] = rawAttributes[ key ];
+					break;
+			}
+		} else {
+			attributes[ key ] = blockAttributes[ key ].default;
+		}
+	} );
+
+	return attributes;
+};
+
+/**
+ * Render Inner Block Components from markup.
+ *
+ * @param {Object}   props           Render props.
+ * @param {Array} props.children  Child nodes.
+ * @param {string}   props.keyPrefix Prefix for keys.
+ * @param {Object}   props.blockMap  Child blocks will be mapped to components.
+ */
+const renderChildren = ( { children, blockMap, keyPrefix = '' } ) => {
```

Comment:
This is where the new magic happens; this converts inner blocks into usable components.